// @datadog/browser-rum/cjs/domain/rumEventsCollection/resource/resourceCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var browser_core_1 = require("@datadog/browser-core");
var types_1 = require("../../../types");
var typesV2_1 = require("../../../typesV2");
var lifeCycle_1 = require("../../lifeCycle");
var matchRequestTiming_1 = require("./matchRequestTiming");
var resourceUtils_1 = require("./resourceUtils");
function startResourceCollection(lifeCycle, configuration, session) {
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.REQUEST_COMPLETED, function (request) {
        if (session.isTrackedWithResource()) {
            configuration.isEnabled('v2_format')
                ? lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, processRequestV2(request))
                : lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processRequest(request));
        }
    });
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.PERFORMANCE_ENTRY_COLLECTED, function (entry) {
        if (session.isTrackedWithResource() && entry.entryType === 'resource' && !resourceUtils_1.isRequestKind(entry)) {
            configuration.isEnabled('v2_format')
                ? lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, processResourceEntryV2(entry))
                : lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processResourceEntry(entry));
        }
    });
}
exports.startResourceCollection = startResourceCollection;
function processRequest(request) {
    var kind = request.type === browser_core_1.RequestType.XHR ? browser_core_1.ResourceType.XHR : browser_core_1.ResourceType.FETCH;
    var matchingTiming = matchRequestTiming_1.matchRequestTiming(request);
    var startTime = matchingTiming ? matchingTiming.startTime : request.startTime;
    var correspondingTimingOverrides = matchingTiming ? computePerformanceEntryMetrics(matchingTiming) : undefined;
    var tracingInfo = computeRequestTracingInfo(request);
    var resourceEvent = browser_core_1.combine({
        date: browser_core_1.getTimestamp(startTime),
        duration: browser_core_1.msToNs(request.duration),
        evt: {
            category: types_1.RumEventCategory.RESOURCE,
        },
        http: {
            method: request.method,
            statusCode: request.status,
            url: request.url,
        },
        resource: {
            kind: kind,
        },
    }, tracingInfo, correspondingTimingOverrides);
    return { startTime: startTime, rawRumEvent: resourceEvent };
}
function processRequestV2(request) {
    var type = request.type === browser_core_1.RequestType.XHR ? browser_core_1.ResourceType.XHR : browser_core_1.ResourceType.FETCH;
    var matchingTiming = matchRequestTiming_1.matchRequestTiming(request);
    var startTime = matchingTiming ? matchingTiming.startTime : request.startTime;
    var correspondingTimingOverrides = matchingTiming ? computePerformanceEntryMetricsV2(matchingTiming) : undefined;
    var tracingInfo = computeRequestTracingInfo(request);
    var resourceEvent = browser_core_1.combine({
        date: browser_core_1.getTimestamp(startTime),
        resource: {
            type: type,
            duration: browser_core_1.msToNs(request.duration),
            method: request.method,
            statusCode: request.status,
            url: request.url,
        },
        type: typesV2_1.RumEventType.RESOURCE,
    }, tracingInfo, correspondingTimingOverrides);
    return { startTime: startTime, rawRumEvent: resourceEvent };
}
function processResourceEntry(entry) {
    var resourceKind = resourceUtils_1.computeResourceKind(entry);
    var entryMetrics = computePerformanceEntryMetrics(entry);
    var tracingInfo = computeEntryTracingInfo(entry);
    var resourceEvent = browser_core_1.combine({
        date: browser_core_1.getTimestamp(entry.startTime),
        evt: {
            category: types_1.RumEventCategory.RESOURCE,
        },
        http: {
            url: entry.name,
        },
        resource: {
            kind: resourceKind,
        },
    }, tracingInfo, entryMetrics);
    return { startTime: entry.startTime, rawRumEvent: resourceEvent };
}
function processResourceEntryV2(entry) {
    var type = resourceUtils_1.computeResourceKind(entry);
    var entryMetrics = computePerformanceEntryMetricsV2(entry);
    var tracingInfo = computeEntryTracingInfo(entry);
    var resourceEvent = browser_core_1.combine({
        date: browser_core_1.getTimestamp(entry.startTime),
        resource: {
            type: type,
            url: entry.name,
        },
        type: typesV2_1.RumEventType.RESOURCE,
    }, tracingInfo, entryMetrics);
    return { startTime: entry.startTime, rawRumEvent: resourceEvent };
}
function computePerformanceEntryMetrics(timing) {
    return {
        duration: resourceUtils_1.computePerformanceResourceDuration(timing),
        http: {
            performance: resourceUtils_1.computePerformanceResourceDetails(timing),
        },
        network: {
            bytesWritten: resourceUtils_1.computeSize(timing),
        },
    };
}
function computePerformanceEntryMetricsV2(timing) {
    return {
        resource: tslib_1.__assign({ duration: resourceUtils_1.computePerformanceResourceDuration(timing), size: resourceUtils_1.computeSize(timing) }, resourceUtils_1.computePerformanceResourceDetails(timing)),
    };
}
function computeRequestTracingInfo(request) {
    var hasBeenTraced = request.traceId && request.spanId;
    if (!hasBeenTraced) {
        return undefined;
    }
    return {
        _dd: {
            spanId: request.spanId.toDecimalString(),
            traceId: request.traceId.toDecimalString(),
        },
        resource: { id: browser_core_1.generateUUID() },
    };
}
function computeEntryTracingInfo(entry) {
    return entry.traceId ? { _dd: { traceId: entry.traceId } } : undefined;
}
