// @datadog/browser-rum/cjs/domain/rumEventsCollection/longTask/longTaskCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_core_1 = require("@datadog/browser-core");
var types_1 = require("../../../types");
var typesV2_1 = require("../../../typesV2");
var lifeCycle_1 = require("../../lifeCycle");
function startLongTaskCollection(lifeCycle, configuration) {
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.PERFORMANCE_ENTRY_COLLECTED, function (entry) {
        if (entry.entryType !== 'longtask') {
            return;
        }
        if (configuration.isEnabled('v2_format')) {
            var rawRumEvent = {
                date: browser_core_1.getTimestamp(entry.startTime),
                longTask: {
                    duration: browser_core_1.msToNs(entry.duration),
                },
                type: typesV2_1.RumEventType.LONG_TASK,
            };
            lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, {
                rawRumEvent: rawRumEvent,
                startTime: entry.startTime,
            });
        }
        else {
            var rawRumEvent = {
                date: browser_core_1.getTimestamp(entry.startTime),
                duration: browser_core_1.msToNs(entry.duration),
                evt: {
                    category: types_1.RumEventCategory.LONG_TASK,
                },
            };
            lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, {
                rawRumEvent: rawRumEvent,
                startTime: entry.startTime,
            });
        }
    });
}
exports.startLongTaskCollection = startLongTaskCollection;
