// @datadog/browser-rum/cjs/boot/rum.entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_core_1 = require("@datadog/browser-core");
var trackActions_1 = require("../domain/rumEventsCollection/action/trackActions");
var rum_1 = require("./rum");
exports.datadogRum = makeRumGlobal(rum_1.startRum);
browser_core_1.defineGlobal(browser_core_1.getGlobalObject(), 'DD_RUM', exports.datadogRum);
function makeRumGlobal(startRumImpl) {
    var isAlreadyInitialized = false;
    var globalContextManager = browser_core_1.createContextManager();
    var getInternalContextStrategy = function () {
        return undefined;
    };
    var beforeInitAddAction = new browser_core_1.BoundedBuffer();
    var addActionStrategy = function (action) {
        beforeInitAddAction.add([action, browser_core_1.deepClone(globalContextManager.get())]);
    };
    var beforeInitAddError = new browser_core_1.BoundedBuffer();
    var addErrorStrategy = function (providedError) {
        beforeInitAddError.add([providedError, browser_core_1.deepClone(globalContextManager.get())]);
    };
    var rumGlobal = browser_core_1.makeGlobal({
        init: browser_core_1.monitor(function (userConfiguration) {
            var _a;
            if (!browser_core_1.checkCookiesAuthorized(browser_core_1.buildCookieOptions(userConfiguration)) ||
                !browser_core_1.checkIsNotLocalFile() ||
                !canInitRum(userConfiguration)) {
                return;
            }
            if (userConfiguration.publicApiKey) {
                userConfiguration.clientToken = userConfiguration.publicApiKey;
            }
            ;
            (_a = startRumImpl(userConfiguration, globalContextManager.get), addActionStrategy = _a.addAction, addErrorStrategy = _a.addError, getInternalContextStrategy = _a.getInternalContext);
            beforeInitAddAction.drain(function (_a) {
                var action = _a[0], context = _a[1];
                return addActionStrategy(action, context);
            });
            beforeInitAddError.drain(function (_a) {
                var error = _a[0], context = _a[1];
                return addErrorStrategy(error, context);
            });
            isAlreadyInitialized = true;
        }),
        addRumGlobalContext: browser_core_1.monitor(globalContextManager.add),
        removeRumGlobalContext: browser_core_1.monitor(globalContextManager.remove),
        getRumGlobalContext: browser_core_1.monitor(globalContextManager.get),
        setRumGlobalContext: browser_core_1.monitor(globalContextManager.set),
        getInternalContext: browser_core_1.monitor(function (startTime) {
            return getInternalContextStrategy(startTime);
        }),
        addAction: browser_core_1.monitor(function (name, context) {
            addActionStrategy({
                name: name,
                context: browser_core_1.deepClone(context),
                startTime: performance.now(),
                type: trackActions_1.ActionType.CUSTOM,
            });
        }),
        addUserAction: function (name, context) {
            // TODO deprecate in v2
            rumGlobal.addAction(name, context);
        },
        addError: browser_core_1.monitor(function (error, context, source) {
            if (source === void 0) { source = browser_core_1.ErrorSource.CUSTOM; }
            var checkedSource;
            if (source === browser_core_1.ErrorSource.CUSTOM || source === browser_core_1.ErrorSource.NETWORK || source === browser_core_1.ErrorSource.SOURCE) {
                checkedSource = source;
            }
            else {
                console.error("DD_RUM.addError: Invalid source '" + source + "'");
                checkedSource = browser_core_1.ErrorSource.CUSTOM;
            }
            addErrorStrategy({
                error: error,
                context: browser_core_1.deepClone(context),
                source: checkedSource,
                startTime: performance.now(),
            });
        }),
    });
    return rumGlobal;
    function canInitRum(userConfiguration) {
        if (isAlreadyInitialized) {
            if (!userConfiguration.silentMultipleInit) {
                console.error('DD_RUM is already initialized.');
            }
            return false;
        }
        if (!userConfiguration || (!userConfiguration.clientToken && !userConfiguration.publicApiKey)) {
            console.error('Client Token is not configured, we will not send any data.');
            return false;
        }
        if (!userConfiguration.applicationId) {
            console.error('Application ID is not configured, no RUM data will be collected.');
            return false;
        }
        if (userConfiguration.sampleRate !== undefined && !browser_core_1.isPercentage(userConfiguration.sampleRate)) {
            console.error('Sample Rate should be a number between 0 and 100');
            return false;
        }
        if (userConfiguration.resourceSampleRate !== undefined && !browser_core_1.isPercentage(userConfiguration.resourceSampleRate)) {
            console.error('Resource Sample Rate should be a number between 0 and 100');
            return false;
        }
        if (Array.isArray(userConfiguration.allowedTracingOrigins) &&
            userConfiguration.allowedTracingOrigins.length !== 0 &&
            userConfiguration.service === undefined) {
            console.error('Service need to be configured when tracing is enabled');
            return false;
        }
        return true;
    }
}
exports.makeRumGlobal = makeRumGlobal;
