// @datadog/browser-rum/cjs/domain/rumEventsCollection/action/actionCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var browser_core_1 = require("@datadog/browser-core");
var types_1 = require("../../../types");
var typesV2_1 = require("../../../typesV2");
var lifeCycle_1 = require("../../lifeCycle");
var trackActions_1 = require("./trackActions");
function startActionCollection(lifeCycle, configuration) {
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.AUTO_ACTION_COMPLETED, function (action) {
        configuration.isEnabled('v2_format')
            ? lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, processActionV2(action))
            : lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, processAction(action));
    });
    if (configuration.trackInteractions) {
        trackActions_1.trackActions(lifeCycle);
    }
    return {
        addAction: function (action, savedGlobalContext) {
            configuration.isEnabled('v2_format')
                ? lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, tslib_1.__assign({ savedGlobalContext: savedGlobalContext }, processActionV2(action)))
                : lifeCycle.notify(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, tslib_1.__assign({ savedGlobalContext: savedGlobalContext }, processAction(action)));
        },
    };
}
exports.startActionCollection = startActionCollection;
function processAction(action) {
    var autoActionProperties = isAutoAction(action)
        ? {
            duration: browser_core_1.msToNs(action.duration),
            userAction: {
                id: action.id,
                measures: action.counts,
            },
        }
        : undefined;
    var customerContext = !isAutoAction(action) ? action.context : undefined;
    var actionEvent = browser_core_1.combine({
        date: browser_core_1.getTimestamp(action.startTime),
        evt: {
            category: types_1.RumEventCategory.USER_ACTION,
            name: action.name,
        },
        userAction: {
            type: action.type,
        },
    }, autoActionProperties);
    return {
        customerContext: customerContext,
        rawRumEvent: actionEvent,
        startTime: action.startTime,
    };
}
function processActionV2(action) {
    var autoActionProperties = isAutoAction(action)
        ? {
            action: {
                error: {
                    count: action.counts.errorCount,
                },
                id: action.id,
                loadingTime: browser_core_1.msToNs(action.duration),
                longTask: {
                    count: action.counts.longTaskCount,
                },
                resource: {
                    count: action.counts.resourceCount,
                },
            },
        }
        : undefined;
    var customerContext = !isAutoAction(action) ? action.context : undefined;
    var actionEvent = browser_core_1.combine({
        action: {
            target: {
                name: action.name,
            },
            type: action.type,
        },
        date: browser_core_1.getTimestamp(action.startTime),
        type: typesV2_1.RumEventType.ACTION,
    }, autoActionProperties);
    return {
        customerContext: customerContext,
        rawRumEvent: actionEvent,
        startTime: action.startTime,
    };
}
function isAutoAction(action) {
    return action.type !== trackActions_1.ActionType.CUSTOM;
}
