// @datadog/browser-rum/cjs/domain/trackEventCounts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_core_1 = require("@datadog/browser-core");
var types_1 = require("../types");
var typesV2_1 = require("../typesV2");
var lifeCycle_1 = require("./lifeCycle");
function trackEventCounts(lifeCycle, callback) {
    if (callback === void 0) { callback = browser_core_1.noop; }
    var eventCounts = {
        errorCount: 0,
        longTaskCount: 0,
        resourceCount: 0,
        userActionCount: 0,
    };
    var subscription = lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_COLLECTED, function (_a) {
        var rawRumEvent = _a.rawRumEvent;
        switch (rawRumEvent.evt.category) {
            case types_1.RumEventCategory.ERROR:
                eventCounts.errorCount += 1;
                callback(eventCounts);
                break;
            case types_1.RumEventCategory.USER_ACTION:
                eventCounts.userActionCount += 1;
                callback(eventCounts);
                break;
            case types_1.RumEventCategory.LONG_TASK:
                eventCounts.longTaskCount += 1;
                callback(eventCounts);
                break;
            case types_1.RumEventCategory.RESOURCE:
                eventCounts.resourceCount += 1;
                callback(eventCounts);
                break;
        }
    });
    var subscriptionV2 = lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.RAW_RUM_EVENT_V2_COLLECTED, function (_a) {
        var rawRumEvent = _a.rawRumEvent;
        switch (rawRumEvent.type) {
            case typesV2_1.RumEventType.ERROR:
                eventCounts.errorCount += 1;
                callback(eventCounts);
                break;
            case typesV2_1.RumEventType.ACTION:
                eventCounts.userActionCount += 1;
                callback(eventCounts);
                break;
            case typesV2_1.RumEventType.LONG_TASK:
                eventCounts.longTaskCount += 1;
                callback(eventCounts);
                break;
            case typesV2_1.RumEventType.RESOURCE:
                eventCounts.resourceCount += 1;
                callback(eventCounts);
                break;
        }
    });
    return {
        stop: function () {
            subscription.unsubscribe();
            subscriptionV2.unsubscribe();
        },
        eventCounts: eventCounts,
    };
}
exports.trackEventCounts = trackEventCounts;
